opentelemetry-sdk/src/opentelemetry/sdk/_metrics/metric_reader_storage.py

```diff
@@ -44,27 +50,60 @@ def _get_or_init_view_instrument_match(
         # Optimistically get the relevant views for the given instrument. Once set for a given
         # instrument, the mapping will never change
 
-        if instrument in self._view_instrument_match:
-            return self._view_instrument_match[instrument]
+        if instrument in self._instrument_view_instrument_matches:
+            return self._instrument_view_instrument_matches[instrument]
 
         with self._lock:
             # double check if it was set before we held the lock
-            if instrument in self._view_instrument_match:
-                return self._view_instrument_match[instrument]
+            if instrument in self._instrument_view_instrument_matches:
+                return self._instrument_view_instrument_matches[instrument]
 
             # not present, hold the lock and add a new mapping
             view_instrument_matches = []
             for view in self._sdk_config.views:
                 # pylint: disable=protected-access
                 if view._match(instrument):
-                    view_instrument_matches.append(
-                        _ViewInstrumentMatch(
-                            view=view,
-                            instrument=instrument,
-                            sdk_config=self._sdk_config,
-                        )
+
+                    new_view_instrument_match = _ViewInstrumentMatch(
+                        view=view,
+                        instrument=instrument,
+                        sdk_config=self._sdk_config,
                     )
 
+                    if isinstance(instrument, Asynchronous) and isinstance(
+                        view._aggregation, ExplicitBucketHistogramAggregation
+                    ):
+                        _logger.warning(
+                            "View %s and instrument %s will produce "
+                            "semantic errors when matched, the view "
+                            "has not been applied.",
+                            view,
+                            instrument,
+                        )
+                        continue
+
+                    for (
+                        existing_view_instrument_matches
+                    ) in self._instrument_view_instrument_matches.values():
+                        for (
+                            existing_view_instrument_match
+                        ) in existing_view_instrument_matches:
+                            if (
+                                existing_view_instrument_match._name
```

Comment:
I am not totally sure about this check. It is included in the spec [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#view):

> Try to apply the View configuration. If applying the View results in conflicting metric identities the implementation SHOULD apply the View and emit a warning.

The conflicting metric identities I can think about should already be stopped by [this exception](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/_metrics/view.py#L110-L118) being raised. We can keep this check if there are other conflicting metric identities you may find.

Reply:
Hmm, I guess what you mean is that in if we have this situation:

``` python
views=[
  View(instrument_name="bar", name="foo"),
  View(instrument_name="baz", name="foo"),
]

# ...
meter.create_counter("foo")
meter.create_counter("bar")
meter.create_counter("baz")
```
... and if the instruments name could match with the view attribute `name` _or_ the view attribute `instrument_name` then we would end up having here 4 `_ViewInstrumentMatch` objects:

* 2 for the instrument named `"foo"` that would match both views
* 1 for the instrument named `"bar"` that would match the view with `instrument_name="bar"`
* 1 for the instrument named `"baz"` that would match the view with `instrument_name="baz"`

and that would cause a conflict when [this condition](https://github.com/open-telemetry/opentelemetry-python/pull/2608/files#diff-4c98dbe0ead9392ad5b77088e9c1393fe59e10ae1edd38c8283170a3bf4b45e5R93-R95) is satisfied, logging a warning.

Nevertheless, that would not happen because the [only](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/_metrics/view.py#L143) view attribute that can match an instrument name is the view `instrument_name` attribute.

So, I think the situation described above can't also cause a conflict. What do you think, @aabmass ?

Reply:
No, i mean this would create 3 output Metric objects in OTLP model with the same identity. We need to log in this case and pass the conflicts through based on the wording in the spec.

Reply:
Ok, I added another test case for the collision situation you mentioned above. I think it is enough to check the `_view_instrument_match._name` attribute when checking for collisions since that is [the name of the metrics stream](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#view) as specified:

> The name of the View (optional). If not provided, the Instrument name MUST be used by default. This will be used as the name of the metrics stream.

Reply:
@aabmass I fixed the matching criteria (added an `__eq__` operator for `_ViewInstrumentMatch` objects) and added more test cases that cover more conflicting and non-conflicting scenarios, PTAL :v:

Reply:
Renamed `__eq__` to `conflicts` since it better represents semantically what the purpose of this function is :+1:

Reply:
Issue mentioned above is fixed, marking this conversation as resolved.